[Merged by Bors] - feat: lint `attribute [instance] instName in`

grunweg opened the pull request:
Contrary to what users would expect, this defines a global instance.

This linter would have caught #13144 and #13170.
[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/ConcreteCategory.2EinstFunLike.20became.20a.20.20global.20instance)

Co-authored by: @adomani

---

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomani reviewed (commented):
This looks good!

I do not know my way around `attribute` syntax, but is the alternative implementation below have some drawbacks?  It seems simpler to me.

Also, could you add a `#guard_msgs`ed test?

```lean
/-- Whether a syntax element is adding an `instance` attribute without a `local` modifier. -/
def is_attribute_instance_in : Syntax → Bool
  | `(command|attribute [instance] $_decl:ident in $_) => true
  | `(command|attribute [instance $_priority] $_decl:ident in $_) => true
  | _ => false

def attributeInstanceIn : Linter where run := withSetOptionIn fun stx => do
  unless getLinterAttributeInstanceIn (← getOptions) do
    return
  if (← MonadState.get).messages.hasErrors then
    return
  if is_attribute_instance_in stx then
    Linter.logLint linter.attributeInstanceIn stx[0] m!
      "careful: `attribute [instance] ... in` declarations are surprising: \
      they are **not** limited to the subsequent declaration, \
      but define global instances \
      please remove the `in` or make this a `local instance` instead"
```

adomani commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
```diff
@@ -0,0 +1,85 @@
+      -- "in" attribute: something non-trivial of index 0
+      match chain.get? 0 with
+      -- This is the case we're interested in
+      | some (_next, 0) => Linter.logLint linter.attributeInstanceIn head m!
+      "careful: `attribute [instance] ... in` declarations are surprising:
```
I do not have an opinion on this, but in `Mathlib` `m!` is never separated from its string.

grunweg replied:
The whole string is way too long for the line, but the `m!` must start on the same line as the `logLint` call... I've tried something in between now.

grunweg commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
```diff
@@ -0,0 +1,85 @@
+      | some (_next, 0) => Linter.logLint linter.attributeInstanceIn head m!
+      "careful: `attribute [instance] ... in` declarations are surprising:
+      they are **not** limited to the subsequent declaration,
+      but define global instances
```
I had not, and was puzzled about the indentation. Thanks, this solves it!

grunweg commented:
I wrote a test yesterday (of course!), but forgot to commit it :facepalm: I've updated it to use your example instance now (and added you as an author of the tests), since this avoid mathlib imports.

grunweg commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
```diff
@@ -0,0 +1,85 @@
+  match stx.findStack? (fun _ ↦ true) is_attribute_instance_in with
+  | some ((head, 0)::chain) =>
+    -- An instance attribute is always the first child of its parent syntax.
+    if is_attribute_instance_in head then
```
Oh, yes, this is also correct.

grunweg commented:
I just tried the logic you proposed. It works, so I switched to it (and added you an an author).
I tweaked one test to consider nested syntax (which worked), so I guess `attribute`s are always the first syntax.

adomani commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
```diff
@@ -0,0 +1,71 @@
+Authors: Michael Rothgang, Damiano Testa
+-/
+
+import Lean.Elab.Command
```
```suggestion
import Lean.Elab.Command
import Lean.Linter.Util
```
I do not know if you were tricked by `Shake`: this import is needed for `logLint`.

grunweg replied:
I was not tricked by shake - I removed this import manually and realised later that it wouldn't build...

adomani reviewed (approved):
I feel like I have had a bit of influence in this PR, so I am going to approve, but will not maintainer merge.

With respect to merging this linter into the "generic" one: I think that this can happen at a later stage, if/when the other PR is more ready.  At the moment, I still consider it experimental.

Thanks!

grunweg commented:
!bench

semorrison commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
````diff
@@ -0,0 +1,72 @@
+*Example*: before this was discovered, `Mathlib/Topology/Category/TopCat/Basic.lean`
+contained the following code:
+```
+-- Porting note: cannot find a coercion to function otherwise
````
```suggestion
```

semorrison commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
```diff
@@ -0,0 +1,72 @@
+    return
+  if is_attribute_instance_in stx then
+    Linter.logLint linter.attributeInstanceIn stx[0] m!
+      "careful: `attribute [instance] ... in` declarations are surprising:\n\
```
Either remove careful, or change to some word already used to introduce linter explanations?

grunweg replied:
Fair point: this linter is not meant to be advisory, but a hard error - I have removed the "careful".
```suggestion
      "`attribute [instance] ... in` declarations are surprising:\n\

semorrison commented:
Two minor requests, otherwise

bors d+

grunweg commented:
Thanks for the fast review.
bors r+

grunweg commented:
bors r+

eric-wieser commented on Mathlib/Tactic/Linter/AttributeInstanceIn.lean:
```diff
@@ -0,0 +1,71 @@
+/-- Whether a syntax element is adding an `instance` attribute without a `local` modifier. -/
+def is_attribute_instance_in (stx : Syntax) : Bool :=
+  match stx with
+  | `(command|attribute [instance] $_decl:ident in $_) => true
```
This should handle the case with multiple idents too

eric-wieser replied:
In fact, it should probably also accept a list of attributes that _contains_ `instance`
